codegen/grid.py

```diff
@@ -0,0 +1,1040 @@
+    param U: the field
+    param i: list of indices (Symbol objects) of field U, e.g. possibley [t,x,y,z] for 3D
+    param k: determine the derivative is with respect to which dimension, e.g. dU/dt if k=0, dU/dx if k=0, assuming i=[t,x,y,z]
+    param d: spacing of the grid, e.g. dx, dt
+    param n: accuracy of approximation, e.g. n=2 for 2nd-order FD approximation
+    returns D=list of expressions for FD approxiation. D[1] for first derivative, D[2] for 2nd derivative etc
+    raises NotImplementedError exception if dimension is more than 4 (1 time and 3 space dimensions)
+    """
+    M = Taylor(d, n)
+    s = [0]*len(i)
+    s[k] = 1 # switch on kth dimension
+    # generate matrix of RHS, i.e. [ ... U[x-1], U[x], U[x+1] ... ]
+    if len(i)==1:
+        RX = Matrix([U[i[0]+s[0]*x] for x in range(-n+1,n)])
+    elif len(i)==2:
+        RX = Matrix([U[i[0]+s[0]*x,i[1]+s[1]*x] for x in range(-n+1,n)])
+    elif len(i)==3:
+        RX = Matrix([U[i[0]+s[0]*x,i[1]+s[1]*x,i[2]+s[2]*x] for x in range(-n+1,n)])
+    elif len(i)==4:
+        RX = Matrix([U[i[0]+s[0]*x,i[1]+s[1]*x,i[2]+s[2]*x,i[3]+s[3]*x] for x in range(-n+1,n)])
+    else:
+        raise NotImplementedError(">4 dimensions, need to fix")
+
+    return M.inv() * RX
+
+def Deriv_half(U, i, k, d, n):
+    """
+    similar function as Deriv() for staggered grids
+    calculate the FD approximation for nth derivative
+    the function works by using M * D = R, as specified in Taylor(), then inverting M, so that D = M_inverse * R
+    param U: the field
+    param i: list of indices (Symbol objects) of field U, e.g. possibley [t,x,y,z] for 3D
+    param k: determine the derivative is with respect to which dimension, e.g. dU/dt if k=0, dU/dx if k=0, assuming i=[t,x,y,z]
+    param d: spacing of the grid, e.g. dx, dt
+    param n: accuracy of approximation, e.g. n=1 for 2nd-order FD approximation, n=2 for 4th-order FD approximation
+    returns D=list of expressions for FD approxiation. D[1] for first derivative, D[2] for 2nd derivative etc
+    raises NotImplementedError exception if dimension is more than 4 (1 time and 3 space dimensions)
+    """
+    M = Taylor_half(d, n)
+    s = [0]*len(i)
+    s[k] = 1 # switch on kth dimension
+    # generate matrix of RHS, i.e. [ ... U[x-1], U[x], U[x+1] ... ]
+    if len(i)==1:
+        RX = Matrix([U[i[0]+s[0]*x*hf] for x in range(-n*2+1,n*2,2)])
+    elif len(i)==2:
+        RX = Matrix([U[i[0]+s[0]*x*hf,i[1]+s[1]*x*hf] for x in range(-n*2+1,n*2,2)])
+    elif len(i)==3:
+        RX = Matrix([U[i[0]+s[0]*x*hf,i[1]+s[1]*x*hf,i[2]+s[2]*x*hf] for x in range(-n*2+1,n*2,2)])
+    elif len(i)==4:
+        RX = Matrix([U[i[0]+s[0]*x*hf,i[1]+s[1]*x*hf,i[2]+s[2]*x*hf,i[3]+s[3]*x*hf] for x in range(-n*2+1,n*2,2)])
+    else:
+        raise NotImplementedError(">4 dimensions, need to fix")
+
+    result = M.inv() * RX
+    return result
+
+def is_half(expr):
+    """
+    test if constants in an expression values to 1/2, 3/2 ...
```

Comment:
that's right.. amended to:
test if constants (non symbols) in an expression is integer+1/2 or integer (indices on staggered grid are either integer or integer+1/2)